pkg/rpcclient/actor/waiter.go

```diff
@@ -0,0 +1,201 @@
+package actor
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/nspcc-dev/neo-go/pkg/core/block"
+	"github.com/nspcc-dev/neo-go/pkg/core/state"
+	"github.com/nspcc-dev/neo-go/pkg/neorpc"
+	"github.com/nspcc-dev/neo-go/pkg/neorpc/result"
+	"github.com/nspcc-dev/neo-go/pkg/rpcclient"
+	"github.com/nspcc-dev/neo-go/pkg/smartcontract/trigger"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+)
+
+var (
+	// ErrTxNotAccepted is returned when transaction wasn't accepted to the chain
+	// even after ValidUntilBlock block persist.
+	ErrTxNotAccepted = errors.New("transaction was not accepted to chain")
+	ErrContextDone   = errors.New("waiter context done")
+)
+
+type (
+	// SimpleWaiter is an interface that enables transaction awaiting functionality
+	// for Actor instance.
+	SimpleWaiter interface {
+		// Context should return the RPC client context to be able to gracefully
+		// shut down all running processes (if so).
+		Context() context.Context
+		GetBlockCount() (uint32, error)
+		GetApplicationLog(hash util.Uint256, trig *trigger.Type) (*result.ApplicationLog, error)
+	}
+	// WSWaiter is an interface that enables improved web-socket based transaction
+	// awaiting functionality.
+	WSWaiter interface {
+		SimpleWaiter
+
+		SubscribeForNewBlocksWithChan(primary *int, rcvrCh chan rpcclient.Notification) (string, error)
+		SubscribeForTransactionExecutionsWithChan(state *string, rcvrCh chan rpcclient.Notification) (string, error)
+		Unsubscribe(id string) error
+	}
+)
+
+// Wait allows to wait until transaction will be accepted to the chain. It can be
+// used as a wrapper for Send or SignAndSend and accepts transaction hash,
+// ValidUntilBlock value and an error. It returns transaction execution result
+// or an error if transaction wasn't accepted to the chain.
+func (a *Actor) Wait(h util.Uint256, vub uint32, err error) (*state.AppExecResult, error) {
+	if err != nil {
+		return nil, err
+	}
+	if wsW, ok := a.client.(WSWaiter); ok {
```

Comment:
How about creating an `actor.PollingWaiter` and an `actor.WSWaiter` that are to be initialized with `RPCPollingWaiter` or `RPCWSWaiter` and `result.Version`? They don't need anything from the `Actor` anyway. Then you could choose and initialize an appropriate one in the `actor.New` and store it internally (or a common `actor.Waiter` interface with `Wait()`).

Reply:
I've implemented it in a separate commit, I'm not sure that everything is OK, so could you check it? And after this I can squash it if needed.